docs/chapter-3/lesson-3.3.md

```diff
@@ -46,24 +46,24 @@ The basic idea behind the solution is as follows:
 - The outer for loop goes through each element in the sequence $2, 3, ..., n$. `i` is the loop variable for this sequence.
 - We begin with the guess that `i` is prime. In code, we do this by setting `flag` to be `True`.
 - Now, we go through all potential divisors of `i`. This is represented by the sequence $2, 3, ..., i - 1$. Variable `j` is the loop variable for this sequence. Notice how the sequence for the inner loop is dependent on `i`, the loop variable for the outer loop.
-- If `j` divides `i`, then `i` cannot be a prime. We correct our initial assumption by updating `flag` to `False` whenever this happens. As we know that `i` is not prime, there is no use of continuing with the inner-loop, so we break out of it.
+- If `j` divides `i` at some point, then `i` cannot be a prime. We correct our initial assumption by updating `flag` to `False` whenever this happens. As we know that `i` is not prime, there is no use of continuing with the inner-loop, so we break out of it.
 - If `j` doesn't divide `i` for any `j` in this sequence, then `i` is a prime. In such a situation, our initial assumption is right, and `flag` stays `True`.
-- Once we are outside the inner-loop, we check if `flag` is `True`. if that is the case, then we increment count as we have hit upon a prime number.
+- Once we are outside the inner-loop, we check if `flag` is `True`. If this is the case, then we increment count as we have hit upon a prime number.
 
 Some important points regarding nested loops:
 
 - Nesting is not restricted to `#!py for` loops. Any one of the following combinations is possible:
-  - `#!py for` inside `#!py for`
-  - `#!py for` inside `#!py while`
-  - `#!py while` inside `#!py while`
-  - `#!py while` inside `#!py for`
+    - `#!py for` inside `#!py for`
+    - `#!py for` inside `#!py while`
+    - `#!py while` inside `#!py while`
+    - `#!py while` inside `#!py for`
 - Multiple levels of nesting is possible.
 
 
 
 ## `#!py while` versus `#!py for`
 
-`#!py for` loops are typically used in situations where the number of iterations can be quantified, whereas `#!py while` loops are used in situations where the number of iterations cannot be quantified exactly. This doesn't mean that the number of iterations in a `#!py for` loop is always constant. For example:
+`#!py for` loops are typically used in situations where the number of iterations can be quantified and is known in advance, whereas `#!py while` loops are used in situations where the number of iterations cannot be quantified exactly. This doesn't mean that the number of iterations in a `#!py for` loop is always constant. For example:
```

Comment:
Actually what I believe is being implied in this paragraph is that in `for` loops, the number of times the code block is run can be determined in advance. Even in your example we know that the body of `for` only needs to executed as much number of times as the input provided. And in case of iterable objects like lists, sets etc the for loop is run exactly as many times as the length of the object.

While on the other is used commonly in cases where we don't have this knowledge in advance.. as you said we just know that it needs to be executed as long as some condition stays `True`.

Many programming tutorials I've seen also introduce the use of `while` and `for` using the same idea. Technicalities and implementation of loops might vary across languages but this basic idea I think applies to all. What say?